### PR TITLE
da65 support for 65816

### PR DESCRIPTION
**Pull request by clbr**
This adds 65816 support to da65. First commit is the compile fix from my previous PR.

### CODE REVIEW

**Comment by mrdudz** [comment]
I am missing two things here:
- some kind of test that verifies the disassembly is correct
- some kind of documentation on that .awk script

**Comment by clbr** [comment]
- there are no existing tests for da65, except 4510
- do you mean outside the header comments? Where?

**Comment by mrdudz** [comment]
> there are no existing tests for da65, except 4510

yes, perhaps - tests are missing all over the place, we have to start somewhere :)

> do you mean outside the header comments? Where?

no, i mean in the user manual - we shouldnt throw random scripts into the tree without explaining them in the manual

**Comment by clbr** [comment]
> no, i mean in the user manual - we shouldnt throw random scripts into the tree without explaining them in the manual

Where in the manual?

**Comment by mrdudz** [comment]
> Where in the manual?

I was going to say "on the target specific page" - but there is no snes target :) Perhaps make a new section "helper scripts" or something like that on the da65 page?

File: src/da65/handler.c [hunk 1]
```diff
@@ -229,6 +233,26 @@ void OH_Immediate (const OpcDesc* D)
...
+        OneLine (D, "#$%04X", GetCodeWord (PC+1));
+    else
+        OneLine (D, "#$%02X", GetCodeByte (PC+1));
+}
+
+
+
+void OH_Immediate65816X (const OpcDesc* D)
+{
+    if (GetAttr (PC) & atIdx16)
```
Comment (mrdudz) [inline src/da65/handler.c:248]: please always use braces with if / else

File: src/da65/infofile.c [hunk 2]
```diff
@@ -637,6 +641,35 @@ static void RangeSection (void)
...
+                InfoNextTok ();
+                InfoAssureStr ();
+                if (InfoSVal[0] == '\0') {
+                    InfoError ("AddrMode may not be empty");
+                }
+                if (InfoSVal[1] == '\0') {
+                    InfoError ("AddrMode must be two characters long");
+                }
+                if (tolower(InfoSVal[0]) == 'm') {
+                    if (InfoSVal[0] == 'm')
```
Comment (mrdudz) [inline src/da65/infofile.c:655]: please always use braces with if / else

File: src/da65/infofile.c [hunk 3]
```diff
@@ -651,6 +684,13 @@ static void RangeSection (void)
         InfoError ("Required values missing from this section");
     }
 
+    if (CPU == CPU_65816) {
+        if (Type == atCode && !(Attributes & tAddrMode))
```
Comment (mrdudz) [inline src/da65/infofile.c:688]: please always use braces with if / else

File: src/da65/labels.c [hunk 4]
```diff
@@ -74,12 +80,26 @@ static const char* MakeLabelName (unsigned Addr)
...
     return LabelBuf;
 }
 
 
 
+static unsigned FindLongIndex (unsigned Addr)
+{
+    unsigned i;
+    for (i = 0; i < LongLabelsUsed; i++) {
+        if (LongSymAddr[i] == Addr)
```
Comment (mrdudz) [inline src/da65/labels.c:94]: please always use braces with if / else

File: src/da65/labels.c [hunk 5]
```diff
@@ -91,19 +111,40 @@ static void AddLabel (unsigned Addr, attr_t Attr, const char* Name)
...
         }
-        Error ("Duplicate label for address $%04X (%s): '%s'", Addr,
-               SymTab[Addr] == 0 ? "<unnamed label>" : SymTab[Addr],
-               Name == 0 ? "<unnamed label>" : Name);
     }
 
     /* Create a new label (xstrdup will return NULL if input NULL) */
-    SymTab[Addr] = xstrdup (Name);
+    if (IsLongAddr (Addr)) {
+        if (LongLabelsUsed >= MAX_LONG_LABELS)
```
Comment (mrdudz) [inline src/da65/labels.c:140]: please always use braces with if / else

File: src/da65/attrtab.c [hunk 6]
```diff
@@ -59,19 +66,36 @@ static unsigned short AttrTab[0x10000];
...
 attr_t GetAttr (unsigned Addr)
 /* Return the attribute for the given address */
 {
     /* Check the given address */
     AddrCheck (Addr);
 
+    if (IsLongAddr (Addr)) {
+        unsigned i;
+        for (i = 0; i < LongAttrsUsed; i++) {
+            if (LongAttrAddr[i] == Addr)
```
Comment (mrdudz) [inline src/da65/attrtab.c:92]: please always use braces with if / else

File: src/da65/attrtab.c [hunk 7]
```diff
@@ -148,6 +172,32 @@ void MarkAddr (unsigned Addr, attr_t Attr)
...
+                        Error ("Duplicate style for long address %06X", Addr);
+                    }
+                }
+                LongAttrVal[i] |= Attr;
+
+                return;
+            }
+        }
+
+        if (LongAttrsUsed >= MAX_LONG_ATTRS)
```
Comment (mrdudz) [inline src/da65/attrtab.c:192]: please always use braces with if / else

File: src/da65/comments.c [hunk 8]
```diff
@@ -60,17 +65,41 @@ static const char* CommentTab[0x10000];
 
 
 
+static unsigned FindLongIndex (unsigned Addr)
+{
+    unsigned i;
+    for (i = 0; i < LongCommentsUsed; i++) {
+        if (LongCommentAddr[i] == Addr)
```
Comment (mrdudz) [inline src/da65/comments.c:72]: please always use braces with if / else

File: src/da65/comments.c [hunk 9]
```diff
@@ -60,17 +65,41 @@ static const char* CommentTab[0x10000];
...
     AddrCheck (Addr);
 
-    /* If we do already have a comment, warn and ignore the new one */
-    if (CommentTab[Addr]) {
-        Warning ("Duplicate comment for address $%04X", Addr);
+    if (IsLongAddr (Addr)) {
+        if (FindLongIndex (Addr)) {
+            Warning ("Duplicate comment for address $%06X", Addr);
+        } else {
+            if (LongCommentsUsed >= MAX_LONG_COMMENTS)
```
Comment (mrdudz) [inline src/da65/comments.c:90]: please always use braces with if / else

File: src/da65/comments.c [hunk 10]
```diff
@@ -82,6 +111,13 @@ const char* GetComment (unsigned Addr)
     /* Check the given address */
     AddrCheck (Addr);
 
+    if (IsLongAddr (Addr)) {
+        const unsigned i = FindLongIndex (Addr);
+        if (i < LongCommentsUsed)
```
Comment (mrdudz) [inline src/da65/comments.c:116]: please always use braces with if / else

File: src/da65/infofile.c [hunk 11]
```diff
@@ -637,6 +641,35 @@ static void RangeSection (void)
...
+                if (tolower(InfoSVal[0]) == 'm') {
+                    if (InfoSVal[0] == 'm')
+                        AddrMode = atMem16;
+                    else
+                        AddrMode = atMem8;
+                } else {
+                    InfoError ("AddrMode syntax: mx");
+                }
+                if (tolower(InfoSVal[1]) == 'x') {
+                    if (InfoSVal[1] == 'x')
```
Comment (mrdudz) [inline src/da65/infofile.c:663]: please always use braces with if / else

**Comment by clbr** [comment]
Braces added.

**Comment by clbr** [comment]
ping

**Comment by mrdudz** [comment]
OK, this has been dangling around long enough :)

**Pull request merged**